src/core/qgsmaplayerserverproperties.h

```diff
@@ -0,0 +1,331 @@
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSMAPLAYERSERVERPROPERTIES_H
+#define QGSMAPLAYERSERVERPROPERTIES_H
+
+#include "qgis_sip.h"
+#include "qgis_core.h"
+#include <QMap>
+#include <QString>
+#include <QMetaType>
+#include <QVariant>
+
+class QgsMapLayer;
+
+class QDomNode;
+class QDomDocument;
+
+/**
+ * \ingroup core
+ * \brief Manages QGIS Server properties for a map layer
+ * \since QGIS 3.22
+ */
+class CORE_EXPORT QgsServerMetadataUrlProperties
+{
+    Q_GADGET
+
+  public:
+
+    /**
+     * \brief MetadataUrl structure.
+     * MetadataUrl is a link to the detailed, standardized metadata about the data.
+     * \ingroup core
+     */
+    struct CORE_EXPORT MetadataUrl
+    {
+
+      /**
+       * Constructor for MetadataUrl.
+       */
+      MetadataUrl( const QString &url = QString(), const QString &type = QString(), const QString &format = QString() )
+        : url( url )
+        , type( type )
+        , format( format )
+      {}
+
+      /**
+       * URL of the link
+       */
+      QString url;
+
+      /**
+       * Link type. Suggested to use FGDC or TC211.
+       */
+      QString type;
+
+      /**
+       * Format specification of online resource. It is strongly suggested to either use text/plain or text/xml.
+       */
+      QString format;
+
+      // TODO c++20 - replace with = default
+
+      //! Compare two MetadataUrl structure.
+      bool operator==( const QgsServerMetadataUrlProperties::MetadataUrl &other ) const;
+    };
+
+    virtual ~QgsServerMetadataUrlProperties() = default;
+
+    /**
+     * Returns a list of metadataUrl resources associated for the layer.
+     * \returns the list of metadata URLs
+     * \see setMetadataUrls()
+     */
+    QList<QgsServerMetadataUrlProperties::MetadataUrl> metadataUrls() const { return mMetadataUrls; };
+
+    /**
+     * Sets a the list of metadata URL for the layer
+     * \see metadataUrls()
+     * \see addMetadataUrl()
+     */
+    void setMetadataUrls( const QList<QgsServerMetadataUrlProperties::MetadataUrl> &metaUrls ) { mMetadataUrls = metaUrls; };
+
+    /**
+     * Add a metadataUrl for the layer
+     * \see setMetadataUrls()
+     */
+    void addMetadataUrl( const QgsServerMetadataUrlProperties::MetadataUrl &metaUrl ) { mMetadataUrls << metaUrl; };
+
+    //! Gets the parent layer
+    virtual const QgsMapLayer *layer() const = 0;
+
+  protected:
+    //! Saves server properties to xml under the layer node
+    void writeXml( QDomNode &layer_node, QDomDocument &document ) const SIP_SKIP;
+
+    //! Reads server properties from project file.
+    void readXml( const QDomNode &layer_node ) SIP_SKIP;
+
+    /**
+     * Copy properties to another instance
+     *
+     * \param properties The properties to copy to
+     */
+    void clone( QgsServerMetadataUrlProperties *properties ) const SIP_SKIP;
```

Comment:
I'm not convinced about the name (a `clone` method generally returns a clone object of the current instance.

Reply:
The `clone` method here is named as is to  follow the same semantic as the `QgsMapLayer::clone` method (where this method is called)

Reply:
Actually the `QgsMapLayer::clone` method is a good example because it returns a clone of the instance (meaning another `QgsMapLayer` object).

So I still think that a `clone` which returns `void` is not very QGIS friendly (from an API point of view).

Reply:
@pblottiere 
> Actually the QgsMapLayer::clone method is a good example because it returns a clone of the instance

I'm confused because the signature of `QgsMapLayer::clone` is
```
void QgsMapLayer::clone( QgsMapLayer *layer ) const
{
```

Reply:
Not the public method: `virtual QgsMapLayer *clone() const = 0;`

Reply:
Well... It's true that your clone method is also protected so I agree to forgot this comment :)

Reply:
Nevermind, I agree that the `clone` name is somewhat confusing. I suggest to rename it to `copyTo` which is closer to what the method really do.

Reply:
Done.

Reply:
Thanks :+1: :)